Add elementHasAttributeNameFilter and elementLocalNameFilter to MutationObserver options

AndrewRot opened the pull request:
Adds additional options when constructing a `MutationObserver`.

MutationObserverInit now contains two additional options: `elementHasAttributeNameFilter` and `elementLocalNameFilter` which are used to enhance the precision of filtering DOM nodes to observe.

See #398 for the background.

Commit:
Add  mutation observer init options elementHasAttributeNameFilter and elementLocalNameFilter

annevk committed:
Meta: update repository files

See https://github.com/whatwg/spec-factory for details.

annevk committed:
Meta: update repository files

See https://github.com/whatwg/spec-factory for details.

domenic committed:
Align with Web IDL extended attribute changes

[TreatNullAs=EmptyString] → [LegacyNullToEmptyString] and [Unforgeable] →
[LegacyUnforgeable]. Follows https://github.com/heycam/webidl/pull/870.

saschanaz committed:
Define ParentNode's replaceChildren()

Tests: https://github.com/web-platform-tests/wpt/pull/22524

Fixes #478. Closes #755 and closes #835.

domenic committed:
Catch errors while upgrading customized built-in elements

Closes https://github.com/whatwg/html/issues/5084.

Tests: https://github.com/web-platform-tests/wpt/pull/23072

annevk committed:
Meta: update repository files

See https://github.com/whatwg/spec-factory for details.

hewholived commented on dom.bs:
```diff
@@ -3437,6 +3450,14 @@ method, when invoked, must run these steps:
  <var>options</var>'s {{MutationObserverInit/characterData}} to true.
 
+ <li><p>If <var>options</var>'s {{MutationObserverInit/elementHasAttributeFilter}} is present and
+ <var>options</var>'s {{MutationObserverInit/childList}} is omitted, then set
```
Remind me why we decided that `childList` should be set to true when either of these options are enabled. I think it's ok to just throw typeError, no?

AndrewRot replied:
It implicitly defaults to checking the immediate node's `childList`. Otherwise it's a bit wonky to have to set  `elementHasAttributeNameFilter` and set `childList` to true every time you want to use `elementHasAttributeNameFilter`

hewholived commented on dom.bs:
```diff
@@ -3594,6 +3623,60 @@ run these steps:
+ {{MutationObserverInit/elementHasAttributeFilter}} is present, then:
+
+ <ol>
+  <li><p>Let <var>nodeAttributesList</var> be <var>nodeAttributesList</var> of the node's attributes, "<code>node.attributes.keys</code>".
```
`node.attribute.keys` may not be accurate. Try to find the appropriate API to capture the attributeNames of a node.

AndrewRot replied:
Possibly `element.getAttributeNames`: https://developer.mozilla.org/en-US/docs/Web/API/Element/getAttributeNames however, this is on `element`, not `node`

Seems like `element.getAttributeNames()` is rather new-ish. Not sure if there is a preference on choosing APIs or code that is _more_ backwards compatible. 

Should I use the handy API `getAttributeNames`, or in the code block, write the polyfill code for it?

hewholived commented on dom.bs:
```diff
@@ -3594,6 +3623,60 @@ run these steps:
+
+      <li><var>options</var>'s {{MutationObserverInit/elementHasAttributeFilter}} is
+      present,and <var>options</var>'s  {{MutationObserverInit/elementHasAttributeFilter}}
+      contains any of the attributes in <var>nodeAttributesList</var>
```
`attributes` ==> `attribute names`

hewholived commented on dom.bs:
```diff
@@ -3398,6 +3398,19 @@ dictionary MutationObserverInit {
    observed and {{MutationObserverInit/attributes}} is true
    or omitted.
+
+   <dt>{{MutationObserverInit/elementHasAttributeFilter}}
```
We are filtering based on just attribute names, right? So, shouldn't it just be `elementHasAttributeNameFilter` ?

hewholived commented on dom.bs:
```diff
@@ -3594,6 +3623,60 @@ run these steps:
+        "<code>node.localName</code>" is in {{MutationObserverInit/elementLocalNameFilter}}
+      <li>if <var>options</var>'s
+        {{MutationObserverInit/elementHasAttributeFilter}} is present,and <var>options</var>'s
+          {{MutationObserverInit/elementHasAttributeFilter}} contains any of the attributes
```
`attribute names`

Yash-Handa committed:
Clarify allowed shadow host local names

Fixes #745.

foolip committed:
Meta: update repository files

See https://github.com/whatwg/spec-factory for details.

foolip committed:
Review Draft Publication: June 2020

shvaikalesh committed:
Editorial: pass invocationTargetInShadowTree to "inner invoke"

foolip committed:
Meta: update repository files

See https://github.com/whatwg/spec-factory for details.

autokagami committed:
Editorial: align with IDL (void → undefined)

annevk committed:
Editorial: use Infra language for Aborting section

annevk committed:
Need to catch exceptions to report them

Now that the default semantics are to throw, we need to be more explicit about catching.

annevk committed:
DOM Events introduction: clarify how listeners are invoked

Fixes #888.

Commit:
minor fixes

annevk committed:
fix formatting

Commit:
minor fixes

mfreed7 committed:
Add "precustomized" custom element state

For https://github.com/whatwg/html/pull/5909.

Tests: https://github.com/web-platform-tests/wpt/pull/25794.

Co-authored-by: Anne van Kesteren <annevk@annevk.nl>

annevk committed:
Meta: link Review Draft

annevk committed:
Meta: put custom element acknowledgment in the IPR section

mfreed7 committed:
Add "is available to element internals"

This is for use with https://github.com/whatwg/html/pull/5912, which
restricts access to ElementInternals's shadowRoot property for shadow
roots which were pre-existing before a custom element
upgrade/construction. See the discussion in
https://github.com/w3c/webcomponents/issues/871#issuecomment-692383222
for more context.

annevk committed:
Editorial: rephrase initEvent() note

Fixes #906.

annevk committed:
Remove Goals and redo Historical section

Now that the relevant W3C documents redirect here, there is no need to explain the differences with them. The Historical section now details what has been removed, without acknowledging the historical organization of those features.

annevk committed:
Correct step in matches()

It needs to pass this rather than an undefined variable. Also clean up the algorithm a bit.

Fixes #921.

chansuke committed:
Correct boundary point comparison in set the start or end

Fixes #924.

benjamingr committed:
Add signal support to addEventListener()

Tests: https://github.com/web-platform-tests/wpt/pull/26472.

Fixes #911.

annevk committed:
Review Draft Publication: December 2020

saschanaz committed:
Use undefined union for window.event

annevk committed:
Meta: default branch rename

Helps with https://github.com/whatwg/meta/issues/174.

annevk committed:
Use a single exception for name validation (follow-up)

Once you have thrown for all QName cases, you can no longer throw for Name cases. Tested as part of #423.

Also cleanup createElementNS()'s domintro.

Fixes #671 and closes #675.

Commit:
add links

sideshowbarker committed:
Editorial: restore necessary comma

This comma got dropped in 83037a112f3f78f071e9c3ad7d979e85dc7e7283 but
the sentence isn’t grammatical without it, so this change restores it.

jasnell committed:
Add AbortSignal.abort() static method

Returns an aborted AbortSignal.

Tests: https://github.com/web-platform-tests/wpt/pull/28003.

Fixes #959.

manekinekko committed:
Editorial: fix typo

annevk committed:
Meta: update repository files

See https://github.com/whatwg/spec-factory for details.

mfreed7 committed:
Add imperative slot assignment API 

Explainer: https://github.com/WICG/webcomponents/blob/gh-pages/proposals/Imperative-Shadow-DOM-Distribution-API.md.

Corresponding HTML PR: https://github.com/whatwg/html/pull/6561.

Closes #860 by superseding it.

Co-authored-by: Yu Han <yuzhehan@yuzhehan-macbookpro.roam.corp.google.com>

annevk committed:
Editorial: major change to align with IDL and Infra

In particular adopt the getter/setter/method/constructor steps pattern.

annevk committed:
Editorial: use modern dictionary syntax

annevk committed:
Meta: remove redundant quotes

Mainly to make the document more consistent which should make it easier to change going forward.

foolip committed:
Add the IDL of XSLTProcessor

Copied from https://wiki.whatwg.org/wiki/DOM_XSLTProcessor.

Part of https://github.com/whatwg/dom/issues/181.

annevk committed:
Editorial: remove context object concept

It has been deprecated for over a year. Now it is obsolete.

foolip committed:
Add ShadowRoot.prototype.delegatesFocus attribute

Tests: https://github.com/web-platform-tests/wpt/pull/28593.

Fixes #931.

annevk committed:
Review Draft Publication: June 2021

lucacasonato committed:
addEventListener(): flatten more returns signal

npm1 committed:
Update Event's timeStamp initialization

Fixes #855.

annevk committed:
Pass eventInitDict to event constructing steps

Fixes #999.

dandclark committed:
Add validity criteria for StaticRange objects

annevk committed:
Revamp the way DOM talks about nodes

1. Address numerous xref issues around the term node and prevent some from occurring for the term children.
2. Embrace Attr as a node even more.
3. Rely on Web IDL's implements and interface primitives to reduce the opportunities for confusion.

Fixes #597, fixes #636, fixes #719, and fixes #770.

AndrewRot committed:
Merge branch 'main' into new-mutation-observer-init-options
